device/sycl/src/seeding/doublet_finding.sycl

```diff
@@ -79,8 +79,10 @@ class DupletFind {
         auto mid_top_doublets_per_bin =
             mid_top_doublet_device.get_items().at(bin_idx);
 
-        auto num_mid_bot_doublets_per_thread = m_localMemBot.get_pointer();
-        auto num_mid_top_doublets_per_thread = m_localMemTop.get_pointer();
+        auto num_mid_bot_doublets_per_thread = m_localMemBot;
+        auto num_mid_top_doublets_per_thread = m_localMemTop;
+        num_mid_bot_doublets_per_thread[workItemIdx] = 0;
+        num_mid_top_doublets_per_thread[workItemIdx] = 0;
```

Comment:
Curious. Notice that you also changed the type of the `num_mid_bot_doublets_per_thread` and `num_mid_top_doublets_per_thread` variables for the rest of the function. :thinking: Which I'm happy with, because using the "SYCL objects" as smart pointers should be safer than to use the raw pointer given back by them.

This does bring up two things though:
  - Is it possible that the initialisation may not even be needed if the manipulation of this local memory is consistently done through the smart pointers?  Though initialising the memory explicitly to `0` certainly seems good.
  - Do you need to make a copy of these objects? I guess you do, since the operator on this functor is `const`. :thinking: I just don't like it all that much that we have to do that copy... But maybe this is indeed the best way to do this.

Reply:
Hmm, yes, the change from calling `get_pointer` sneaked in here. And just removing that didn't solve the problem, the `num_mid_bot_doublets_per_thread` and `num_mid_top_doublets_per_thread` had to be initialized to `0`. 
But that means that right now, the types of those arrays are:
```c++
 sycl::accessor<int, 1, sycl::access::mode::read_write,::sycl::access::target::local>;
```
whereas before (after calling `get_pointer`), they were of the type
```c++
sycl::multi_ptr<int, :sycl::access::target::local>
```
So is `sycl::accessor` equivalent to a smart pointer in this case and `sycl::multi_ptr` to the raw pointer?

It seems like I do make copies of those objects since as you've said, the `operator ()` is marked `const`, but I'm not really sure why the compiler doesn't allow me to change that :thinking:

Reply:
:smile: It's part of the SYCL standard that the "executable" that we send to the device must be immutable. If it wasn't, if it was allowed to modify values held by the class, you could get the idea that threads would be allowed to communicate with each other through those values.

The member variables of these functors are basically behaving like arguments to a CUDA kernel. But while a CUDA compiler needs to invent its own language to handle global functions with a variadic list of arguments, SYCL kernels pass a variadic list of arguments to the kernel through such member variables using bog standard C\+\+.

But let's not delay it any longer. The code looks good like this.